# DATA 607 -- Assignment 3

In this assignment, we apply the ideas underlying dense word embeddings like Word2Vec and GloVe to construct dense embeddings of categorical features.

The context of our exploration will be the [Rossmann Store Sales Competition](https://www.kaggle.com/c/rossmann-store-sales/overview/description) from *Kaggle*, the goal of which is to forecast store sales using store, promotion, and competitor data.

## Instructions

1. Download the data from the competition page or from [my github](https://github.com/mgreenbe/rossmann).

2. Replace each date in the `Date` column with number of days between it and January 1, 2013, the earliest date in the table.

3. Use `pd.get_dummies` to construct dataframes `stores`, `days_of_week`, and `state_holidays` containing 1-hot encodings of the categorical variables `Store`, `DayOfWeek`, and `StateHoliday`, respectively.

4. Assemble these encoded features, together with the numerical ones (`Date`, `Customers`) and binary ones (`Open`, `Promo`, `SchoolHoliday`), in a matrix `X`, the first 1115 columns of which represent the store ID.

5. Split the data `X` and `Y` into training and validation sets. Standardize the numerical feature columns. Here, the relevant means and standard deviations should be computed from *training data*.

6. Train the model `MyModel`, below, using `MeanSquaredLogarithmicError` as the loss function. Explain, briefly, why this is an appropriate choice of loss function. Stop training when validation error stabilizes.

7. **(Optional)** Add hidden layers to this model and tune the `store_emb_dim` hyperparameter to improve your results.


In [85]:
from tensorflow import keras

class MyModel(keras.Model):
  def __init__(self, n_stores=1115, store_emb_dim=20):
    super(MyModel, self).__init__()
    self.n_stores = n_stores
    self.encoder = keras.layers.Dense(store_emb_dim, name="encoder")
    self.regressor = keras.layers.Dense(1, name="regressor")

  def call(self, X):
    x = tf.concat([self.encoder(X[:, :self.n_stores]), X[:, self.n_stores:]], axis=-1)
    return self.regressor(x)

- This is model is *not* built with `keras.models.Sequential` -- it's not simply passing data through a sequence of layers. The first 1115 columns of the input, representing the store ID, are projected onto a `store_emb_dim`-dimensional space. The resulting projections are then concatenated with the remaining features before applying linear regression. (Notice the absence of nonlinear activation functions.)

- **Warning:** The data set contains > 1 million rows. To avoid running out of memory, work initially with a subset of the rows (say, 10,000). Train on as large a subset of the whole dataset as you can without crashing your session.

## References

Rachel Thomas, [An Introduction to Deep Learning for Tabular Data](https://www.fast.ai/2018/04/29/categorical-embeddings/) (fast.ai blog, April 29, 2018)

Cheng Guo and Felix Berkhahn, [Entity Embeddings of Categorical Variables](https://arxiv.org/pdf/1604.06737.pdf) (April 25, 2016)

In [86]:
import pandas as pd
import numpy as np
from datetime import date
from datetime import datetime
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from keras.callbacks import *

## Q1. 

In [87]:
test_url='https://raw.githubusercontent.com/mgreenbe/rossmann/main/test.csv'
train_url='https://raw.githubusercontent.com/mgreenbe/rossmann/main/train.csv'

In [88]:
df_test = pd.read_csv(test_url,encoding="latin-1")
df_test.head()

Id  Store  DayOfWeek        Date  Open  Promo StateHoliday  SchoolHoliday
0   1      1          4  2015-09-17   1.0      1            0              0
1   2      3          4  2015-09-17   1.0      1            0              0
2   3      7          4  2015-09-17   1.0      1            0              0
3   4      8          4  2015-09-17   1.0      1            0              0
4   5      9          4  2015-09-17   1.0      1            0              0

In [89]:
df_test.shape

(41088, 8)

In [90]:
df_train = pd.read_csv(train_url,encoding="latin-1",dtype=object)
df_train.head()

Store DayOfWeek        Date  Sales Customers Open Promo StateHoliday  \
0     1         5  2015-07-31   5263       555    1     1            0   
1     2         5  2015-07-31   6064       625    1     1            0   
2     3         5  2015-07-31   8314       821    1     1            0   
3     4         5  2015-07-31  13995      1498    1     1            0   
4     5         5  2015-07-31   4822       559    1     1            0   

  SchoolHoliday  
0             1  
1             1  
2             1  
3             1  
4             1

In [91]:
df_train.shape

(1017209, 9)

## Q2.

In [92]:
#changing the date format to what the actual data is like
date_format = "%Y-%m-%d"
#earliest date 
earliest_date = datetime.strptime('2013-01-01', date_format)

for rows in range(len(df_train)):
    #obtaining the date at current row of traning dataframe
    current_date=datetime.strptime(df_train.iat[rows,2], date_format)
    #replacing date
    delta=current_date - earliest_date
    df_train.iat[rows,2]= delta.days

In [93]:
df_train.head()

Store DayOfWeek Date  Sales Customers Open Promo StateHoliday SchoolHoliday
0     1         5  941   5263       555    1     1            0             1
1     2         5  941   6064       625    1     1            0             1
2     3         5  941   8314       821    1     1            0             1
3     4         5  941  13995      1498    1     1            0             1
4     5         5  941   4822       559    1     1            0             1

In [94]:
#same logic as above but for test data
for rows in range(len(df_test)):
    #obtaining the date at current row of traning dataframe
    current_date=datetime.strptime(str(df_test.iat[rows,3]), date_format)
    #replacing date
    delta=current_date - earliest_date
    df_test.iat[rows,3]= delta.days


In [95]:
df_test.head()

Id  Store  DayOfWeek Date  Open  Promo StateHoliday  SchoolHoliday
0   1      1          4  989   1.0      1            0              0
1   2      3          4  989   1.0      1            0              0
2   3      7          4  989   1.0      1            0              0
3   4      8          4  989   1.0      1            0              0
4   5      9          4  989   1.0      1            0              0

## Q3.

In [96]:
#for training data
store_train =pd.get_dummies(df_train.Store)
days_of_week_train =pd.get_dummies(df_train.DayOfWeek)
state_holidays_train=pd.get_dummies(df_train.StateHoliday)

In [97]:
len(df_train.Store.unique())

1115

In [98]:
store_train.shape

(1017209, 1115)

In [99]:
store_train.head(1)

1  10  100  1000  1001  1002  1003  1004  1005  1006  ...  990  991  992  \
0  1   0    0     0     0     0     0     0     0     0  ...    0    0    0   

   993  994  995  996  997  998  999  
0    0    0    0    0    0    0    0  

[1 rows x 1115 columns]

In [100]:
days_of_week_train.head(1)

1  2  3  4  5  6  7
0  0  0  0  0  1  0  0

In [101]:
days_of_week_train.shape

(1017209, 7)

It can be observed that the number of columns in store_train dataframe are the unique entries of the Store column in df_train dataset

In [102]:
#for test data
store_test =pd.get_dummies(df_test.Store)
days_of_week_test =pd.get_dummies(df_test.DayOfWeek)
state_holidays_test=pd.get_dummies(df_test.StateHoliday)

## Q4.

In [103]:
#train
X=pd.concat([store_train, days_of_week_train,state_holidays_train,df_train.Date,df_train.Sales,df_train.Customers,df_train.Open,
            df_train.Promo,df_train.SchoolHoliday], axis=1)

In [104]:
X.shape

(1017209, 1132)

In [105]:
#test
Y=pd.concat([store_test, days_of_week_test,state_holidays_test,df_test.Date,df_test.Open,
            df_test.Promo,df_test.SchoolHoliday], axis=1)

In [106]:
Y.shape

(41088, 869)

## Q5.

Since the test data does not have the Sales variable I will not use it.I will instead explcitly state Y is sales. Further, I will drop the Sales column from the X dataframe.

In [107]:
Y=X.Sales

In [108]:
X.drop(columns=['Sales'],inplace=True)

In [144]:
X_train,X_val,Y_train,Y_val=train_test_split(X, Y, test_size=0.2, random_state=42)

**Standadizing numerical feature columns(Date ,Customers for X)
with means and standard deviation obtaining in the training data**

In [145]:
X_train.shape, X_val.shape, Y_train.shape, Y_val.shape

((813767, 1131), (203442, 1131), (813767,), (203442,))

In [146]:
scaler_Date = StandardScaler().fit(df_train.Date.values.reshape(-1,1))
scaler_Customer = StandardScaler().fit(df_train.Customers.values.reshape(-1,1))

In [138]:
print(scaler_Date.mean_,scaler_Customer.mean_)

[465.0629959] [633.14594641]


**Standardize data**

In [139]:
X_train.head(1)

1  10  100  1000  1001  1002  1003  1004  1005  1006  ...  7  0  a  b  \
417683  0   0    0     0     0     0     0     0     0     0  ...  1  1  0  0   

        c  Date  Customers  Open  Promo  SchoolHoliday  
417683  0   537          0     0      0              0  

[1 rows x 1131 columns]

In [140]:
X_val.head(1)

1  10  100  1000  1001  1002  1003  1004  1005  1006  ...  7  0  a  b  \
76435  0   0    0     0     0     0     0     0     0     0  ...  1  1  0  0   

       c  Date  Customers  Open  Promo  SchoolHoliday  
76435  0   873          0     0      0              0  

[1 rows x 1131 columns]

**Standardize Date**

In [147]:
X_train.Date=scaler_Date.transform(X_train.Date.values.reshape(-1,1))
X_val.Date=scaler_Date.transform(X_val.Date.values.reshape(-1,1))

C:\Users\Edwin Aguirre\AppData\Roaming\Python\Python37\site-packages\pandas\core\generic.py:5491: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value
C:\Users\Edwin Aguirre\AppData\Roaming\Python\Python37\site-packages\pandas\core\generic.py:5491: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


**Standardize Customers**

In [148]:
X_train.Customers=scaler_Customer.transform(X_train.Customers.values.reshape(-1,1))
X_val.Customers=scaler_Customer.transform(X_val.Customers.values.reshape(-1,1))

C:\Users\Edwin Aguirre\AppData\Roaming\Python\Python37\site-packages\pandas\core\generic.py:5491: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value
C:\Users\Edwin Aguirre\AppData\Roaming\Python\Python37\site-packages\pandas\core\generic.py:5491: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [149]:
X_train1.head(1)

1  10  100  1000  1001  1002  1003  1004  1005  1006  ...  7  0  a  b  \
459018  0   0    0     0     0     0     0     0     0     0  ...  0  1  0  0   

        c      Date  Customers  Open  Promo  SchoolHoliday  
459018  0  0.127296   0.197786     1      0              0  

[1 rows x 1131 columns]

In [150]:
X_val1.head(1)

1  10  100  1000  1001  1002  1003  1004  1005  1006  ...  7  0  a  b  \
816535  0   0    0     0     0     0     0     0     0     0  ...  0  1  0  0   

        c      Date  Customers  Open  Promo  SchoolHoliday  
816535  0 -1.042299   -1.36333     1      0              0  

[1 rows x 1131 columns]

## Q6.

**Model Defined**

In [151]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.layers import Dense, Embedding,GRU
class MyModel(keras.Model):
    def __init__(self, n_stores=1115, store_emb_dim=20):
        super(MyModel, self).__init__()
        self.n_stores = n_stores
        self.encoder = keras.layers.Dense(store_emb_dim, name="encoder")
        self.regressor = keras.layers.Dense(1, name="regressor")

    def call(self, X):
        x = tf.concat([self.encoder(X[:, :self.n_stores]), X[:, self.n_stores:]], axis=-1)
        return self.regressor(x)

**Training model**

#### Random sampling of 20,000 to skip out of memory issue

In [155]:
X_validamall=X_val.sample(n=20000, random_state=42)
Y_validamall=Y_val[X_validamall.index]

In [156]:
X_trainsmall=X_train.sample(n=20000, random_state=42)
Y_trainsmall=Y_train[X_trainsmall.index]
#Adding callbacks
cb = EarlyStopping(monitor='val_mean_squared_logarithmic_error', mode='auto', verbose=1,patience=10)  

In [157]:
X_trainsmall.shape, Y_trainsmall.shape, X_validamall.shape,Y_validamall.shape

((20000, 1131), (20000,), (20000, 1131), (20000,))

In [158]:
model=MyModel()
model.compile(optimizer='adam',loss="MeanSquaredLogarithmicError",metrics=["MeanSquaredLogarithmicError"])
model.fit(x=np.array(X_trainsmall,dtype=np.float16),
          y=np.array(Y_trainsmall,dtype=np.float16),
          validation_data=(np.array(X_validamall,dtype=np.float16),np.array(Y_validamall,dtype=np.float16)),
          epochs=10000,callbacks=[cb])

Epoch 1/10000
625/625 [==============================] - 2s 2ms/step - loss: 51.5471 - mean_squared_logarithmic_error: 51.5471 - val_loss: 33.7653 - val_mean_squared_logarithmic_error: 33.7653
Epoch 2/10000
625/625 [==============================] - 1s 1ms/step - loss: 31.5602 - mean_squared_logarithmic_error: 31.5602 - val_loss: 26.6889 - val_mean_squared_logarithmic_error: 26.6889
Epoch 3/10000
625/625 [==============================] - 1s 1ms/step - loss: 25.5834 - mean_squared_logarithmic_error: 25.5834 - val_loss: 22.8969 - val_mean_squared_logarithmic_error: 22.8969
Epoch 4/10000
625/625 [==============================] - 1s 1ms/step - loss: 22.2294 - mean_squared_logarithmic_error: 22.2294 - val_loss: 20.3830 - val_mean_squared_logarithmic_error: 20.3830
Epoch 5/10000
625/625 [==============================] - 1s 1ms/step - loss: 19.8651 - mean_squared_logarithmic_error: 19.8651 - val_loss: 18.5408 - val_mean_squared_logarithmic_error: 18.5408
Epoch 6/10000
625/625 [============

625/625 [==============================] - 1s 1ms/step - loss: 9.4465 - mean_squared_logarithmic_error: 9.4465 - val_loss: 10.6819 - val_mean_squared_logarithmic_error: 10.6819
Epoch 86/10000
625/625 [==============================] - 1s 2ms/step - loss: 9.4985 - mean_squared_logarithmic_error: 9.4985 - val_loss: 10.6711 - val_mean_squared_logarithmic_error: 10.6711
Epoch 87/10000
625/625 [==============================] - 1s 2ms/step - loss: 9.2967 - mean_squared_logarithmic_error: 9.2967 - val_loss: 10.6482 - val_mean_squared_logarithmic_error: 10.6482
Epoch 88/10000
625/625 [==============================] - 1s 2ms/step - loss: 9.4257 - mean_squared_logarithmic_error: 9.4257 - val_loss: 10.6388 - val_mean_squared_logarithmic_error: 10.6388
Epoch 89/10000
625/625 [==============================] - 1s 2ms/step - loss: 9.3295 - mean_squared_logarithmic_error: 9.3295 - val_loss: 10.6214 - val_mean_squared_logarithmic_error: 10.6214
Epoch 90/10000
625/625 [==============================]

Epoch 170/10000
625/625 [==============================] - 1s 2ms/step - loss: 6.9634 - mean_squared_logarithmic_error: 6.9634 - val_loss: 9.2954 - val_mean_squared_logarithmic_error: 9.2954
Epoch 171/10000
625/625 [==============================] - 1s 2ms/step - loss: 7.0378 - mean_squared_logarithmic_error: 7.0378 - val_loss: 9.2807 - val_mean_squared_logarithmic_error: 9.2807
Epoch 172/10000
625/625 [==============================] - 1s 2ms/step - loss: 7.0649 - mean_squared_logarithmic_error: 7.0649 - val_loss: 9.2654 - val_mean_squared_logarithmic_error: 9.2654
Epoch 173/10000
625/625 [==============================] - 1s 2ms/step - loss: 7.0434 - mean_squared_logarithmic_error: 7.0434 - val_loss: 9.2534 - val_mean_squared_logarithmic_error: 9.2534
Epoch 174/10000
625/625 [==============================] - 1s 2ms/step - loss: 7.0092 - mean_squared_logarithmic_error: 7.0092 - val_loss: 9.2386 - val_mean_squared_logarithmic_error: 9.2386
Epoch 175/10000
625/625 [====================

Epoch 213/10000
625/625 [==============================] - 1s 2ms/step - loss: 6.2644 - mean_squared_logarithmic_error: 6.2644 - val_loss: 8.6366 - val_mean_squared_logarithmic_error: 8.6366
Epoch 214/10000
625/625 [==============================] - 1s 2ms/step - loss: 6.3536 - mean_squared_logarithmic_error: 6.3536 - val_loss: 8.6214 - val_mean_squared_logarithmic_error: 8.6214
Epoch 215/10000
625/625 [==============================] - 1s 2ms/step - loss: 6.1721 - mean_squared_logarithmic_error: 6.1721 - val_loss: 8.6092 - val_mean_squared_logarithmic_error: 8.6092
Epoch 216/10000
625/625 [==============================] - 1s 2ms/step - loss: 6.2376 - mean_squared_logarithmic_error: 6.2376 - val_loss: 8.5911 - val_mean_squared_logarithmic_error: 8.5911
Epoch 217/10000
625/625 [==============================] - 1s 1ms/step - loss: 6.0666 - mean_squared_logarithmic_error: 6.0666 - val_loss: 8.5769 - val_mean_squared_logarithmic_error: 8.5769
Epoch 218/10000
625/625 [====================

Epoch 256/10000
625/625 [==============================] - 1s 2ms/step - loss: 5.6556 - mean_squared_logarithmic_error: 5.6556 - val_loss: 7.9921 - val_mean_squared_logarithmic_error: 7.9921
Epoch 257/10000
625/625 [==============================] - 1s 2ms/step - loss: 5.6566 - mean_squared_logarithmic_error: 5.6566 - val_loss: 7.9758 - val_mean_squared_logarithmic_error: 7.9758
Epoch 258/10000
625/625 [==============================] - 1s 1ms/step - loss: 5.6265 - mean_squared_logarithmic_error: 5.6265 - val_loss: 7.9628 - val_mean_squared_logarithmic_error: 7.9628
Epoch 259/10000
625/625 [==============================] - 1s 2ms/step - loss: 5.6016 - mean_squared_logarithmic_error: 5.6016 - val_loss: 7.9464 - val_mean_squared_logarithmic_error: 7.9464
Epoch 260/10000
625/625 [==============================] - 1s 1ms/step - loss: 5.6864 - mean_squared_logarithmic_error: 5.6864 - val_loss: 7.9367 - val_mean_squared_logarithmic_error: 7.9367
Epoch 261/10000
625/625 [====================

Epoch 299/10000
625/625 [==============================] - 1s 2ms/step - loss: 5.1604 - mean_squared_logarithmic_error: 5.1604 - val_loss: 7.4347 - val_mean_squared_logarithmic_error: 7.4347
Epoch 300/10000
625/625 [==============================] - 1s 2ms/step - loss: 5.2355 - mean_squared_logarithmic_error: 5.2355 - val_loss: 7.4210 - val_mean_squared_logarithmic_error: 7.4210
Epoch 301/10000
625/625 [==============================] - 1s 2ms/step - loss: 5.1227 - mean_squared_logarithmic_error: 5.1227 - val_loss: 7.4069 - val_mean_squared_logarithmic_error: 7.4069
Epoch 302/10000
625/625 [==============================] - 1s 2ms/step - loss: 5.0462 - mean_squared_logarithmic_error: 5.0462 - val_loss: 7.3957 - val_mean_squared_logarithmic_error: 7.3957
Epoch 303/10000
625/625 [==============================] - 1s 2ms/step - loss: 5.0102 - mean_squared_logarithmic_error: 5.0102 - val_loss: 7.3834 - val_mean_squared_logarithmic_error: 7.3834
Epoch 304/10000
625/625 [====================

Epoch 342/10000
625/625 [==============================] - 1s 2ms/step - loss: 4.6373 - mean_squared_logarithmic_error: 4.6373 - val_loss: 6.8883 - val_mean_squared_logarithmic_error: 6.8883
Epoch 343/10000
625/625 [==============================] - 1s 2ms/step - loss: 4.6797 - mean_squared_logarithmic_error: 4.6797 - val_loss: 6.8787 - val_mean_squared_logarithmic_error: 6.8787
Epoch 344/10000
625/625 [==============================] - 1s 1ms/step - loss: 4.6280 - mean_squared_logarithmic_error: 4.6280 - val_loss: 6.8738 - val_mean_squared_logarithmic_error: 6.8738
Epoch 345/10000
625/625 [==============================] - 1s 2ms/step - loss: 4.5389 - mean_squared_logarithmic_error: 4.5389 - val_loss: 6.8626 - val_mean_squared_logarithmic_error: 6.8626
Epoch 346/10000
625/625 [==============================] - 1s 2ms/step - loss: 4.7341 - mean_squared_logarithmic_error: 4.7341 - val_loss: 6.8534 - val_mean_squared_logarithmic_error: 6.8534
Epoch 347/10000
625/625 [====================

Epoch 385/10000
625/625 [==============================] - 2s 3ms/step - loss: 4.2967 - mean_squared_logarithmic_error: 4.2967 - val_loss: 6.4954 - val_mean_squared_logarithmic_error: 6.4954
Epoch 386/10000
625/625 [==============================] - 1s 2ms/step - loss: 4.2305 - mean_squared_logarithmic_error: 4.2305 - val_loss: 6.4879 - val_mean_squared_logarithmic_error: 6.4879
Epoch 387/10000
625/625 [==============================] - 2s 2ms/step - loss: 4.3173 - mean_squared_logarithmic_error: 4.3173 - val_loss: 6.4805 - val_mean_squared_logarithmic_error: 6.4805
Epoch 388/10000
625/625 [==============================] - 1s 2ms/step - loss: 4.1749 - mean_squared_logarithmic_error: 4.1749 - val_loss: 6.4736 - val_mean_squared_logarithmic_error: 6.4736
Epoch 389/10000
625/625 [==============================] - 2s 3ms/step - loss: 4.2509 - mean_squared_logarithmic_error: 4.2509 - val_loss: 6.4655 - val_mean_squared_logarithmic_error: 6.4655
Epoch 390/10000
625/625 [====================

Epoch 428/10000
625/625 [==============================] - 1s 2ms/step - loss: 3.9717 - mean_squared_logarithmic_error: 3.9717 - val_loss: 6.1019 - val_mean_squared_logarithmic_error: 6.1019
Epoch 429/10000
625/625 [==============================] - 1s 2ms/step - loss: 3.9682 - mean_squared_logarithmic_error: 3.9682 - val_loss: 6.0927 - val_mean_squared_logarithmic_error: 6.0927
Epoch 430/10000
625/625 [==============================] - 1s 2ms/step - loss: 3.9699 - mean_squared_logarithmic_error: 3.9699 - val_loss: 6.0841 - val_mean_squared_logarithmic_error: 6.0841
Epoch 431/10000
625/625 [==============================] - 1s 2ms/step - loss: 3.8912 - mean_squared_logarithmic_error: 3.8912 - val_loss: 6.0731 - val_mean_squared_logarithmic_error: 6.0731
Epoch 432/10000
625/625 [==============================] - 1s 2ms/step - loss: 3.9504 - mean_squared_logarithmic_error: 3.9504 - val_loss: 6.0678 - val_mean_squared_logarithmic_error: 6.0678
Epoch 433/10000
625/625 [====================

Epoch 471/10000
625/625 [==============================] - 1s 1ms/step - loss: 3.6613 - mean_squared_logarithmic_error: 3.6613 - val_loss: 5.6932 - val_mean_squared_logarithmic_error: 5.6932
Epoch 472/10000
625/625 [==============================] - 1s 1ms/step - loss: 3.5974 - mean_squared_logarithmic_error: 3.5974 - val_loss: 5.6891 - val_mean_squared_logarithmic_error: 5.6891
Epoch 473/10000
625/625 [==============================] - 1s 1ms/step - loss: 3.5551 - mean_squared_logarithmic_error: 3.5551 - val_loss: 5.6736 - val_mean_squared_logarithmic_error: 5.6736
Epoch 474/10000
625/625 [==============================] - 1s 1ms/step - loss: 3.6563 - mean_squared_logarithmic_error: 3.6563 - val_loss: 5.6689 - val_mean_squared_logarithmic_error: 5.6689
Epoch 475/10000
625/625 [==============================] - 1s 1ms/step - loss: 3.6130 - mean_squared_logarithmic_error: 3.6130 - val_loss: 5.6570 - val_mean_squared_logarithmic_error: 5.6570
Epoch 476/10000
625/625 [====================

Epoch 514/10000
625/625 [==============================] - 1s 1ms/step - loss: 3.3806 - mean_squared_logarithmic_error: 3.3806 - val_loss: 5.3284 - val_mean_squared_logarithmic_error: 5.3284
Epoch 515/10000
625/625 [==============================] - 1s 1ms/step - loss: 3.3514 - mean_squared_logarithmic_error: 3.3514 - val_loss: 5.3196 - val_mean_squared_logarithmic_error: 5.3196
Epoch 516/10000
625/625 [==============================] - 1s 2ms/step - loss: 3.3817 - mean_squared_logarithmic_error: 3.3817 - val_loss: 5.3104 - val_mean_squared_logarithmic_error: 5.3104
Epoch 517/10000
625/625 [==============================] - 1s 1ms/step - loss: 3.3056 - mean_squared_logarithmic_error: 3.3056 - val_loss: 5.3010 - val_mean_squared_logarithmic_error: 5.3010
Epoch 518/10000
625/625 [==============================] - 1s 1ms/step - loss: 3.4297 - mean_squared_logarithmic_error: 3.4297 - val_loss: 5.2970 - val_mean_squared_logarithmic_error: 5.2970
Epoch 519/10000
625/625 [====================

Epoch 557/10000
625/625 [==============================] - 1s 1ms/step - loss: 3.0941 - mean_squared_logarithmic_error: 3.0941 - val_loss: 5.0076 - val_mean_squared_logarithmic_error: 5.0076
Epoch 558/10000
625/625 [==============================] - 1s 1ms/step - loss: 3.1657 - mean_squared_logarithmic_error: 3.1657 - val_loss: 4.9955 - val_mean_squared_logarithmic_error: 4.9955
Epoch 559/10000
625/625 [==============================] - 1s 2ms/step - loss: 3.1806 - mean_squared_logarithmic_error: 3.1806 - val_loss: 4.9857 - val_mean_squared_logarithmic_error: 4.9857
Epoch 560/10000
625/625 [==============================] - 1s 2ms/step - loss: 3.2043 - mean_squared_logarithmic_error: 3.2043 - val_loss: 4.9792 - val_mean_squared_logarithmic_error: 4.9792
Epoch 561/10000
625/625 [==============================] - 1s 2ms/step - loss: 3.1603 - mean_squared_logarithmic_error: 3.1603 - val_loss: 4.9722 - val_mean_squared_logarithmic_error: 4.9722
Epoch 562/10000
625/625 [====================

Epoch 600/10000
625/625 [==============================] - 1s 1ms/step - loss: 2.9356 - mean_squared_logarithmic_error: 2.9356 - val_loss: 4.7013 - val_mean_squared_logarithmic_error: 4.7013
Epoch 601/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.9144 - mean_squared_logarithmic_error: 2.9144 - val_loss: 4.6931 - val_mean_squared_logarithmic_error: 4.6931
Epoch 602/10000
625/625 [==============================] - 1s 1ms/step - loss: 2.9281 - mean_squared_logarithmic_error: 2.9281 - val_loss: 4.6879 - val_mean_squared_logarithmic_error: 4.6879
Epoch 603/10000
625/625 [==============================] - 1s 1ms/step - loss: 2.9637 - mean_squared_logarithmic_error: 2.9637 - val_loss: 4.6801 - val_mean_squared_logarithmic_error: 4.6801
Epoch 604/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.9652 - mean_squared_logarithmic_error: 2.9652 - val_loss: 4.6803 - val_mean_squared_logarithmic_error: 4.6803
Epoch 605/10000
625/625 [====================

Epoch 643/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.7535 - mean_squared_logarithmic_error: 2.7535 - val_loss: 4.4252 - val_mean_squared_logarithmic_error: 4.4252
Epoch 644/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.7492 - mean_squared_logarithmic_error: 2.7492 - val_loss: 4.4184 - val_mean_squared_logarithmic_error: 4.4184
Epoch 645/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.7657 - mean_squared_logarithmic_error: 2.7657 - val_loss: 4.4122 - val_mean_squared_logarithmic_error: 4.4122
Epoch 646/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.7536 - mean_squared_logarithmic_error: 2.7536 - val_loss: 4.4084 - val_mean_squared_logarithmic_error: 4.4084
Epoch 647/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.7837 - mean_squared_logarithmic_error: 2.7837 - val_loss: 4.4000 - val_mean_squared_logarithmic_error: 4.4000
Epoch 648/10000
625/625 [====================

Epoch 686/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.5794 - mean_squared_logarithmic_error: 2.5794 - val_loss: 4.1622 - val_mean_squared_logarithmic_error: 4.1622
Epoch 687/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.6209 - mean_squared_logarithmic_error: 2.6209 - val_loss: 4.1590 - val_mean_squared_logarithmic_error: 4.1590
Epoch 688/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.6119 - mean_squared_logarithmic_error: 2.6119 - val_loss: 4.1514 - val_mean_squared_logarithmic_error: 4.1514
Epoch 689/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.5837 - mean_squared_logarithmic_error: 2.5837 - val_loss: 4.1459 - val_mean_squared_logarithmic_error: 4.1459
Epoch 690/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.5807 - mean_squared_logarithmic_error: 2.5807 - val_loss: 4.1374 - val_mean_squared_logarithmic_error: 4.1374
Epoch 691/10000
625/625 [====================

Epoch 729/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.4527 - mean_squared_logarithmic_error: 2.4527 - val_loss: 3.9251 - val_mean_squared_logarithmic_error: 3.9251
Epoch 730/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.4517 - mean_squared_logarithmic_error: 2.4517 - val_loss: 3.9164 - val_mean_squared_logarithmic_error: 3.9164
Epoch 731/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.4579 - mean_squared_logarithmic_error: 2.4579 - val_loss: 3.9091 - val_mean_squared_logarithmic_error: 3.9091
Epoch 732/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.4329 - mean_squared_logarithmic_error: 2.4329 - val_loss: 3.9065 - val_mean_squared_logarithmic_error: 3.9065
Epoch 733/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.4891 - mean_squared_logarithmic_error: 2.4891 - val_loss: 3.9000 - val_mean_squared_logarithmic_error: 3.9000
Epoch 734/10000
625/625 [====================

Epoch 772/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.2869 - mean_squared_logarithmic_error: 2.2869 - val_loss: 3.7077 - val_mean_squared_logarithmic_error: 3.7077
Epoch 773/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.3540 - mean_squared_logarithmic_error: 2.3540 - val_loss: 3.7033 - val_mean_squared_logarithmic_error: 3.7033
Epoch 774/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.3397 - mean_squared_logarithmic_error: 2.3397 - val_loss: 3.6994 - val_mean_squared_logarithmic_error: 3.6994
Epoch 775/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.3377 - mean_squared_logarithmic_error: 2.3377 - val_loss: 3.6961 - val_mean_squared_logarithmic_error: 3.6961
Epoch 776/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.3060 - mean_squared_logarithmic_error: 2.3060 - val_loss: 3.6912 - val_mean_squared_logarithmic_error: 3.6912
Epoch 777/10000
625/625 [====================

Epoch 815/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.2129 - mean_squared_logarithmic_error: 2.2129 - val_loss: 3.5153 - val_mean_squared_logarithmic_error: 3.5153
Epoch 816/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.2074 - mean_squared_logarithmic_error: 2.2074 - val_loss: 3.5114 - val_mean_squared_logarithmic_error: 3.5114
Epoch 817/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.1725 - mean_squared_logarithmic_error: 2.1725 - val_loss: 3.5054 - val_mean_squared_logarithmic_error: 3.5054
Epoch 818/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.2135 - mean_squared_logarithmic_error: 2.2135 - val_loss: 3.5009 - val_mean_squared_logarithmic_error: 3.5009
Epoch 819/10000
625/625 [==============================] - 1s 1ms/step - loss: 2.1925 - mean_squared_logarithmic_error: 2.1925 - val_loss: 3.5000 - val_mean_squared_logarithmic_error: 3.5000
Epoch 820/10000
625/625 [====================

Epoch 858/10000
625/625 [==============================] - 1s 1ms/step - loss: 2.1173 - mean_squared_logarithmic_error: 2.1173 - val_loss: 3.3159 - val_mean_squared_logarithmic_error: 3.3159
Epoch 859/10000
625/625 [==============================] - 1s 1ms/step - loss: 2.0965 - mean_squared_logarithmic_error: 2.0965 - val_loss: 3.3145 - val_mean_squared_logarithmic_error: 3.3145
Epoch 860/10000
625/625 [==============================] - 1s 1ms/step - loss: 2.1091 - mean_squared_logarithmic_error: 2.1091 - val_loss: 3.3169 - val_mean_squared_logarithmic_error: 3.3169
Epoch 861/10000
625/625 [==============================] - 1s 1ms/step - loss: 2.0671 - mean_squared_logarithmic_error: 2.0671 - val_loss: 3.3063 - val_mean_squared_logarithmic_error: 3.3063
Epoch 862/10000
625/625 [==============================] - 1s 1ms/step - loss: 2.0407 - mean_squared_logarithmic_error: 2.0407 - val_loss: 3.3050 - val_mean_squared_logarithmic_error: 3.3050
Epoch 863/10000
625/625 [====================

Epoch 901/10000
625/625 [==============================] - 1s 2ms/step - loss: 1.9911 - mean_squared_logarithmic_error: 1.9911 - val_loss: 3.1440 - val_mean_squared_logarithmic_error: 3.1440
Epoch 902/10000
625/625 [==============================] - 1s 2ms/step - loss: 2.0163 - mean_squared_logarithmic_error: 2.0163 - val_loss: 3.1396 - val_mean_squared_logarithmic_error: 3.1396
Epoch 903/10000
625/625 [==============================] - 1s 2ms/step - loss: 1.9909 - mean_squared_logarithmic_error: 1.9909 - val_loss: 3.1349 - val_mean_squared_logarithmic_error: 3.1349
Epoch 904/10000
625/625 [==============================] - 1s 2ms/step - loss: 1.9682 - mean_squared_logarithmic_error: 1.9682 - val_loss: 3.1332 - val_mean_squared_logarithmic_error: 3.1332
Epoch 905/10000
625/625 [==============================] - 1s 2ms/step - loss: 1.9905 - mean_squared_logarithmic_error: 1.9905 - val_loss: 3.1296 - val_mean_squared_logarithmic_error: 3.1296
Epoch 906/10000
625/625 [====================

Epoch 944/10000
625/625 [==============================] - 1s 2ms/step - loss: 1.8977 - mean_squared_logarithmic_error: 1.8977 - val_loss: 2.9822 - val_mean_squared_logarithmic_error: 2.9822
Epoch 945/10000
625/625 [==============================] - 1s 2ms/step - loss: 1.8890 - mean_squared_logarithmic_error: 1.8890 - val_loss: 2.9828 - val_mean_squared_logarithmic_error: 2.9828
Epoch 946/10000
625/625 [==============================] - 1s 2ms/step - loss: 1.8892 - mean_squared_logarithmic_error: 1.8892 - val_loss: 2.9780 - val_mean_squared_logarithmic_error: 2.9780
Epoch 947/10000
625/625 [==============================] - 1s 2ms/step - loss: 1.8869 - mean_squared_logarithmic_error: 1.8869 - val_loss: 2.9727 - val_mean_squared_logarithmic_error: 2.9727
Epoch 948/10000
625/625 [==============================] - 1s 2ms/step - loss: 1.8872 - mean_squared_logarithmic_error: 1.8872 - val_loss: 2.9742 - val_mean_squared_logarithmic_error: 2.9742
Epoch 949/10000
625/625 [====================

Epoch 987/10000
625/625 [==============================] - 1s 2ms/step - loss: 1.8021 - mean_squared_logarithmic_error: 1.8021 - val_loss: 2.8711 - val_mean_squared_logarithmic_error: 2.8711
Epoch 988/10000
625/625 [==============================] - 1s 2ms/step - loss: 1.8151 - mean_squared_logarithmic_error: 1.8151 - val_loss: 2.8698 - val_mean_squared_logarithmic_error: 2.8698
Epoch 989/10000
625/625 [==============================] - 2s 3ms/step - loss: 1.7964 - mean_squared_logarithmic_error: 1.7964 - val_loss: 2.8642 - val_mean_squared_logarithmic_error: 2.8642
Epoch 990/10000
625/625 [==============================] - 2s 3ms/step - loss: 1.8173 - mean_squared_logarithmic_error: 1.8173 - val_loss: 2.8615 - val_mean_squared_logarithmic_error: 2.8615
Epoch 991/10000
625/625 [==============================] - 1s 2ms/step - loss: 1.8078 - mean_squared_logarithmic_error: 1.8078 - val_loss: 2.8625 - val_mean_squared_logarithmic_error: 2.8625
Epoch 992/10000
625/625 [====================

Epoch 1030/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.7181 - mean_squared_logarithmic_error: 1.7181 - val_loss: 2.7548 - val_mean_squared_logarithmic_error: 2.7548
Epoch 1031/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.7229 - mean_squared_logarithmic_error: 1.7229 - val_loss: 2.7506 - val_mean_squared_logarithmic_error: 2.7506
Epoch 1032/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.7098 - mean_squared_logarithmic_error: 1.7098 - val_loss: 2.7505 - val_mean_squared_logarithmic_error: 2.7505
Epoch 1033/10000
625/625 [==============================] - 1s 2ms/step - loss: 1.7114 - mean_squared_logarithmic_error: 1.7114 - val_loss: 2.7495 - val_mean_squared_logarithmic_error: 2.7495
Epoch 1034/10000
625/625 [==============================] - 1s 2ms/step - loss: 1.7068 - mean_squared_logarithmic_error: 1.7068 - val_loss: 2.7465 - val_mean_squared_logarithmic_error: 2.7465
Epoch 1035/10000
625/625 [==============

Epoch 1073/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.6291 - mean_squared_logarithmic_error: 1.6291 - val_loss: 2.6584 - val_mean_squared_logarithmic_error: 2.6584
Epoch 1074/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.6405 - mean_squared_logarithmic_error: 1.6405 - val_loss: 2.6621 - val_mean_squared_logarithmic_error: 2.6621
Epoch 1075/10000
625/625 [==============================] - 1s 2ms/step - loss: 1.6424 - mean_squared_logarithmic_error: 1.6424 - val_loss: 2.6588 - val_mean_squared_logarithmic_error: 2.6588
Epoch 1076/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.6395 - mean_squared_logarithmic_error: 1.6395 - val_loss: 2.6555 - val_mean_squared_logarithmic_error: 2.6555
Epoch 1077/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.6352 - mean_squared_logarithmic_error: 1.6352 - val_loss: 2.6553 - val_mean_squared_logarithmic_error: 2.6553
Epoch 1078/10000
625/625 [==============

Epoch 1116/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.5691 - mean_squared_logarithmic_error: 1.5691 - val_loss: 2.5868 - val_mean_squared_logarithmic_error: 2.5868
Epoch 1117/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.5754 - mean_squared_logarithmic_error: 1.5754 - val_loss: 2.5838 - val_mean_squared_logarithmic_error: 2.5838
Epoch 1118/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.5673 - mean_squared_logarithmic_error: 1.5673 - val_loss: 2.5865 - val_mean_squared_logarithmic_error: 2.5865
Epoch 1119/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.5661 - mean_squared_logarithmic_error: 1.5661 - val_loss: 2.5817 - val_mean_squared_logarithmic_error: 2.5817
Epoch 1120/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.5791 - mean_squared_logarithmic_error: 1.5791 - val_loss: 2.5823 - val_mean_squared_logarithmic_error: 2.5823
Epoch 1121/10000
625/625 [==============

Epoch 1159/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.5264 - mean_squared_logarithmic_error: 1.5264 - val_loss: 2.5199 - val_mean_squared_logarithmic_error: 2.5199
Epoch 1160/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.5151 - mean_squared_logarithmic_error: 1.5151 - val_loss: 2.5175 - val_mean_squared_logarithmic_error: 2.5175
Epoch 1161/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.5099 - mean_squared_logarithmic_error: 1.5099 - val_loss: 2.5174 - val_mean_squared_logarithmic_error: 2.5174
Epoch 1162/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.5096 - mean_squared_logarithmic_error: 1.5096 - val_loss: 2.5210 - val_mean_squared_logarithmic_error: 2.5210
Epoch 1163/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.4956 - mean_squared_logarithmic_error: 1.4956 - val_loss: 2.5147 - val_mean_squared_logarithmic_error: 2.5147
Epoch 1164/10000
625/625 [==============

Epoch 1202/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.4601 - mean_squared_logarithmic_error: 1.4601 - val_loss: 2.4581 - val_mean_squared_logarithmic_error: 2.4581
Epoch 1203/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.4632 - mean_squared_logarithmic_error: 1.4632 - val_loss: 2.4545 - val_mean_squared_logarithmic_error: 2.4545
Epoch 1204/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.4449 - mean_squared_logarithmic_error: 1.4449 - val_loss: 2.4541 - val_mean_squared_logarithmic_error: 2.4541
Epoch 1205/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.4491 - mean_squared_logarithmic_error: 1.4491 - val_loss: 2.4551 - val_mean_squared_logarithmic_error: 2.4551
Epoch 1206/10000
625/625 [==============================] - 1s 2ms/step - loss: 1.4378 - mean_squared_logarithmic_error: 1.4378 - val_loss: 2.4510 - val_mean_squared_logarithmic_error: 2.4510
Epoch 1207/10000
625/625 [==============

Epoch 1245/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.4089 - mean_squared_logarithmic_error: 1.4089 - val_loss: 2.3961 - val_mean_squared_logarithmic_error: 2.3961
Epoch 1246/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.4013 - mean_squared_logarithmic_error: 1.4013 - val_loss: 2.3929 - val_mean_squared_logarithmic_error: 2.3929
Epoch 1247/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.4071 - mean_squared_logarithmic_error: 1.4071 - val_loss: 2.3922 - val_mean_squared_logarithmic_error: 2.3922
Epoch 1248/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.4028 - mean_squared_logarithmic_error: 1.4028 - val_loss: 2.3907 - val_mean_squared_logarithmic_error: 2.3907
Epoch 1249/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.3988 - mean_squared_logarithmic_error: 1.3988 - val_loss: 2.3903 - val_mean_squared_logarithmic_error: 2.3903
Epoch 1250/10000
625/625 [==============

Epoch 1288/10000
625/625 [==============================] - 1s 2ms/step - loss: 1.3499 - mean_squared_logarithmic_error: 1.3499 - val_loss: 2.3427 - val_mean_squared_logarithmic_error: 2.3427
Epoch 1289/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.3425 - mean_squared_logarithmic_error: 1.3425 - val_loss: 2.3442 - val_mean_squared_logarithmic_error: 2.3442
Epoch 1290/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.3445 - mean_squared_logarithmic_error: 1.3445 - val_loss: 2.3393 - val_mean_squared_logarithmic_error: 2.3393
Epoch 1291/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.3494 - mean_squared_logarithmic_error: 1.3494 - val_loss: 2.3370 - val_mean_squared_logarithmic_error: 2.3370
Epoch 1292/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.3428 - mean_squared_logarithmic_error: 1.3428 - val_loss: 2.3334 - val_mean_squared_logarithmic_error: 2.3334
Epoch 1293/10000
625/625 [==============

Epoch 1331/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.3071 - mean_squared_logarithmic_error: 1.3071 - val_loss: 2.2849 - val_mean_squared_logarithmic_error: 2.2849
Epoch 1332/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.3065 - mean_squared_logarithmic_error: 1.3065 - val_loss: 2.2875 - val_mean_squared_logarithmic_error: 2.2875
Epoch 1333/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.2879 - mean_squared_logarithmic_error: 1.2879 - val_loss: 2.2850 - val_mean_squared_logarithmic_error: 2.2850
Epoch 1334/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.3027 - mean_squared_logarithmic_error: 1.3027 - val_loss: 2.2847 - val_mean_squared_logarithmic_error: 2.2847
Epoch 1335/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.3059 - mean_squared_logarithmic_error: 1.3059 - val_loss: 2.2835 - val_mean_squared_logarithmic_error: 2.2835
Epoch 1336/10000
625/625 [==============

Epoch 1374/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.2446 - mean_squared_logarithmic_error: 1.2446 - val_loss: 2.2412 - val_mean_squared_logarithmic_error: 2.2412
Epoch 1375/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.2656 - mean_squared_logarithmic_error: 1.2656 - val_loss: 2.2314 - val_mean_squared_logarithmic_error: 2.2314
Epoch 1376/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.2440 - mean_squared_logarithmic_error: 1.2440 - val_loss: 2.2318 - val_mean_squared_logarithmic_error: 2.2318
Epoch 1377/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.2487 - mean_squared_logarithmic_error: 1.2487 - val_loss: 2.2319 - val_mean_squared_logarithmic_error: 2.2319
Epoch 1378/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.2568 - mean_squared_logarithmic_error: 1.2568 - val_loss: 2.2308 - val_mean_squared_logarithmic_error: 2.2308
Epoch 1379/10000
625/625 [==============

Epoch 1417/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.2045 - mean_squared_logarithmic_error: 1.2045 - val_loss: 2.1887 - val_mean_squared_logarithmic_error: 2.1887
Epoch 1418/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.2095 - mean_squared_logarithmic_error: 1.2095 - val_loss: 2.1947 - val_mean_squared_logarithmic_error: 2.1947
Epoch 1419/10000
625/625 [==============================] - 1s 2ms/step - loss: 1.2033 - mean_squared_logarithmic_error: 1.2033 - val_loss: 2.1889 - val_mean_squared_logarithmic_error: 2.1889
Epoch 1420/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.2118 - mean_squared_logarithmic_error: 1.2118 - val_loss: 2.1875 - val_mean_squared_logarithmic_error: 2.1875
Epoch 1421/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.1967 - mean_squared_logarithmic_error: 1.1967 - val_loss: 2.1804 - val_mean_squared_logarithmic_error: 2.1804
Epoch 1422/10000
625/625 [==============

Epoch 1460/10000
625/625 [==============================] - 1s 2ms/step - loss: 1.1646 - mean_squared_logarithmic_error: 1.1646 - val_loss: 2.1379 - val_mean_squared_logarithmic_error: 2.1379
Epoch 1461/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.1626 - mean_squared_logarithmic_error: 1.1626 - val_loss: 2.1399 - val_mean_squared_logarithmic_error: 2.1399
Epoch 1462/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.1596 - mean_squared_logarithmic_error: 1.1596 - val_loss: 2.1377 - val_mean_squared_logarithmic_error: 2.1377
Epoch 1463/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.1461 - mean_squared_logarithmic_error: 1.1461 - val_loss: 2.1391 - val_mean_squared_logarithmic_error: 2.1391
Epoch 1464/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.1643 - mean_squared_logarithmic_error: 1.1643 - val_loss: 2.1366 - val_mean_squared_logarithmic_error: 2.1366
Epoch 1465/10000
625/625 [==============

Epoch 1503/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.1230 - mean_squared_logarithmic_error: 1.1230 - val_loss: 2.0964 - val_mean_squared_logarithmic_error: 2.0964
Epoch 1504/10000
625/625 [==============================] - 1s 2ms/step - loss: 1.1278 - mean_squared_logarithmic_error: 1.1278 - val_loss: 2.0952 - val_mean_squared_logarithmic_error: 2.0952
Epoch 1505/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.1275 - mean_squared_logarithmic_error: 1.1275 - val_loss: 2.0970 - val_mean_squared_logarithmic_error: 2.0970
Epoch 1506/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.1374 - mean_squared_logarithmic_error: 1.1374 - val_loss: 2.1032 - val_mean_squared_logarithmic_error: 2.1032
Epoch 1507/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.1269 - mean_squared_logarithmic_error: 1.1269 - val_loss: 2.0900 - val_mean_squared_logarithmic_error: 2.0900
Epoch 1508/10000
625/625 [==============

Epoch 1546/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.0904 - mean_squared_logarithmic_error: 1.0904 - val_loss: 2.0512 - val_mean_squared_logarithmic_error: 2.0512
Epoch 1547/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.0888 - mean_squared_logarithmic_error: 1.0888 - val_loss: 2.0565 - val_mean_squared_logarithmic_error: 2.0565
Epoch 1548/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.0942 - mean_squared_logarithmic_error: 1.0942 - val_loss: 2.0513 - val_mean_squared_logarithmic_error: 2.0513
Epoch 1549/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.0908 - mean_squared_logarithmic_error: 1.0908 - val_loss: 2.0537 - val_mean_squared_logarithmic_error: 2.0537
Epoch 1550/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.0861 - mean_squared_logarithmic_error: 1.0861 - val_loss: 2.0540 - val_mean_squared_logarithmic_error: 2.0540
Epoch 1551/10000
625/625 [==============

Epoch 1589/10000
625/625 [==============================] - 1s 2ms/step - loss: 1.0546 - mean_squared_logarithmic_error: 1.0546 - val_loss: 2.0213 - val_mean_squared_logarithmic_error: 2.0213
Epoch 1590/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.0543 - mean_squared_logarithmic_error: 1.0543 - val_loss: 2.0168 - val_mean_squared_logarithmic_error: 2.0168
Epoch 1591/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.0491 - mean_squared_logarithmic_error: 1.0491 - val_loss: 2.0194 - val_mean_squared_logarithmic_error: 2.0194
Epoch 1592/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.0481 - mean_squared_logarithmic_error: 1.0481 - val_loss: 2.0135 - val_mean_squared_logarithmic_error: 2.0135
Epoch 1593/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.0488 - mean_squared_logarithmic_error: 1.0488 - val_loss: 2.0164 - val_mean_squared_logarithmic_error: 2.0164
Epoch 1594/10000
625/625 [==============

Epoch 1632/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.0202 - mean_squared_logarithmic_error: 1.0202 - val_loss: 1.9807 - val_mean_squared_logarithmic_error: 1.9807
Epoch 1633/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.0105 - mean_squared_logarithmic_error: 1.0105 - val_loss: 1.9769 - val_mean_squared_logarithmic_error: 1.9769
Epoch 1634/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.0098 - mean_squared_logarithmic_error: 1.0098 - val_loss: 1.9704 - val_mean_squared_logarithmic_error: 1.9704
Epoch 1635/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.0139 - mean_squared_logarithmic_error: 1.0139 - val_loss: 1.9812 - val_mean_squared_logarithmic_error: 1.9812
Epoch 1636/10000
625/625 [==============================] - 1s 1ms/step - loss: 1.0137 - mean_squared_logarithmic_error: 1.0137 - val_loss: 1.9776 - val_mean_squared_logarithmic_error: 1.9776
Epoch 1637/10000
625/625 [==============

Epoch 1675/10000
625/625 [==============================] - 1s 1ms/step - loss: 0.9888 - mean_squared_logarithmic_error: 0.9888 - val_loss: 1.9435 - val_mean_squared_logarithmic_error: 1.9435
Epoch 1676/10000
625/625 [==============================] - 1s 1ms/step - loss: 0.9930 - mean_squared_logarithmic_error: 0.9930 - val_loss: 1.9407 - val_mean_squared_logarithmic_error: 1.9407
Epoch 1677/10000
625/625 [==============================] - 1s 1ms/step - loss: 0.9969 - mean_squared_logarithmic_error: 0.9969 - val_loss: 1.9381 - val_mean_squared_logarithmic_error: 1.9381
Epoch 1678/10000
625/625 [==============================] - 1s 1ms/step - loss: 0.9822 - mean_squared_logarithmic_error: 0.9822 - val_loss: 1.9409 - val_mean_squared_logarithmic_error: 1.9409
Epoch 1679/10000
625/625 [==============================] - 1s 1ms/step - loss: 0.9835 - mean_squared_logarithmic_error: 0.9835 - val_loss: 1.9360 - val_mean_squared_logarithmic_error: 1.9360
Epoch 1680/10000
625/625 [==============

After 1682 epochs, the validation error stabilized. Therefore, the training of the model "MyModel" stopped. 

**Why is using MSLE a good idea?**

In [173]:
scaler_Y = StandardScaler().fit(Y.values.reshape(-1,1))

In [174]:
print(scaler_Y.mean_,scaler_Y.var_)

[5773.81897231] [14821916.98358157]


Using MSLE is a good idea when the predictor variable has large deviations (high variance),and you don’t want large errors to be significantly more penalized than small ones.Since, the variance of our predictor is quite large(~14821917), using the MSLE is a good idea.